app/services/oauth_clients/connection_manager.rb

```diff
@@ -88,10 +91,57 @@ def code_to_token(code)
       service_result = request_new_token(authorization_code: code)
       return service_result unless service_result.success?
 
-      # Create a new OAuthClientToken from Rack::OAuth::AccessToken::Bearer and return
-      ServiceResult.success(
-        result: create_new_oauth_client_token(service_result.result)
+      # Check for existing OAuthClientToken and update,
+      # or create a new one from Rack::OAuth::AccessToken::Bearer
+      oauth_client_token = get_existing_token
+      if oauth_client_token.present?
+        update_oauth_client_token(oauth_client_token, service_result.result)
+      else
+        oauth_client_token = create_new_oauth_client_token(service_result.result)
+      end
+
+      ServiceResult.success(result: oauth_client_token)
+    end
+
+    # Called by StorageRepresenter to inquire about the status of the OAuth2
+    # authentication server.
+    # Returns :connected/:authorization_failed or :error for a general error.
+    # We have decided to distinguish between only these 3 cases, because the
+    # front-end (and a normal user) probably wouldn't know how to deal with
+    # other options.
+    def authorization_state
+      oauth_client_token = get_existing_token
+      return :failed_authorization unless oauth_client_token
+
+      # Check for user information. This is the cheapest Nextcloud call that requires
+      # valid authentication, so we use it for testing the validity of the Bearer token.
+      # curl -H "Authorization: Bearer MY_TOKEN" -X GET 'https://my.nextcloud.org/ocs/v1.php/cloud/user' \
+      #      -H "OCS-APIRequest: true" -H "Accept: application/json"
+      RestClient.get(
+        File.join(oauth_client.integration.host, AUTHORIZATION_CHECK_PATH),
+        {
+          'Authorization' => "Bearer #{oauth_client_token.access_token}",
+          'OCS-APIRequest' => "true",
+          'Accept' => "application/json"
+        }
       )
```

Comment:
The implementation now becomes Nextcloud specific (url & headers). This could be solved by extracting the checks into strategies per storage type. But it can also be done later, once different types exist.

Reply:
If okay, I would postpone this because of all the other things that need to be done.